File: src/leetcode/array/TopKFrequent.java
package leetcode.array;

import java.util.*;

/**
 * Given a non-empty array of integers, return the k most frequent elements.
 * Example:
 * Input: nums = [1,1,1,2,2,3], k = 2
 * Output: [1,2]
 */
public class TopKFrequent {

    public int[] topKFrequentBruteForce(int[] nums, int k) {
        if (nums.length == k) {
            return nums;
        }

        // Create a map to store the frequent counts of the elements
        Map<Integer, Integer> frequentCounts = new HashMap<>();
        for (int num : nums) {
            // Update the frequent count of the element
            frequentCounts.put(num, frequentCounts.getOrDefault(num, 0) + 1);
        }

        // Initialize the result array with size k
        int[] result = new int[k];
        // Iterate over the frequent counts and find the top k frequent elements
        for (int i = 0; i < k; i++) {
            // Initialize the max count and key
            int maxCount = 0;
            int maxKey = 0;

            // Find the maximum frequent element
            for (Map.Entry<Integer, Integer> entry : frequentCounts.entrySet()) {
                // Update the max count and key
                if (entry.getValue() > maxCount) {
                    maxCount = entry.getValue();
                    maxKey = entry.getKey();
                }
            }

            // Add the max element to the result
            result[i] = maxKey;
            // Remove the max element from the frequent counts
            frequentCounts.remove(maxKey);
        }

        return result;
    }

    public int[] topKFrequentOptimal(int[] nums, int k) {
        if (nums.length == k) {
            return nums;
        }

        Map<Integer, Integer> numberCounts = new HashMap<>();
        Map<Integer, List<Integer>> countFrequencies = new HashMap<>();

        // Count the frequency of each element
        for (int num : nums) {
            numberCounts.put(num, numberCounts.getOrDefault(num, 0) + 1);
        }

        // Store the elements with the same frequency in a list
        for (Map.Entry<Integer, Integer> entry : numberCounts.entrySet()) {
            if (!countFrequencies.containsKey(entry.getValue())) {
                countFrequencies.put(entry.getValue(), new ArrayList<>());
            }

            countFrequencies.get(entry.getValue()).add(entry.getKey());
        }

        // Initialize the result array with size k
        int[] result = new int[k];
        int resultIndex = 0;
        // Iterate over the frequencies in descending order
        for (int i = nums.length; i > 0; i--) {
            if (countFrequencies.containsKey(i)) {
                // Add the elements with the same frequency to the result
                for (int num : countFrequencies.get(i)) {
                    result[resultIndex++] = num;
                    // Return the result if the k elements are found
                    if (resultIndex == k) {
                        return result;
                    }
                }
            }
        }

        return result;
    }

    public static void main(String[] args) {
        TopKFrequent kFrequent = new TopKFrequent();
        int[] bruteForceResult = kFrequent.topKFrequentBruteForce(new int[]{1, 1, 1, 2, 2, 5}, 2);
        int[] optimalResult = kFrequent.topKFrequentOptimal(new int[]{4, 1, -1, 2, -1, 2, 3}, 2);
        System.out.println(Arrays.toString(bruteForceResult));
        System.out.println(Arrays.toString(optimalResult));
    }
}

File: src/leetcode/array/top_k_frequent.py
class Solution:
    """
     Given a non-empty array of integers, return the k most frequent elements.
     Example:
     Input: nums = [1,1,1,2,2,3], k = 2
     Output: [1,2]
    """

    def topKFrequentBruteForce(self, nums, k):
        if len(nums) == k:
            return nums

        # Count the frequency of each element
        frequent_counts = {}
        for num in nums:
            # Update the count of the element
            frequent_counts[num] = frequent_counts.get(num, 0) + 1

        # Initialize the result list
        result = list(range(k))
        for i in range(k):
            # Initialize the max count and key
            max_count = 0
            max_key = 0

            # Find the key with the maximum count
            for key, value in frequent_counts.items():
                # Update the max count and key
                if value > max_count:
                    max_count = value
                    max_key = key

            # Update the result list
            result[i] = max_key
            # Remove the key from the dictionary
            del frequent_counts[max_key]

        return result
    
    def kTopFrequentOptimal(self, nums, k):
        number_counts = {}
        # Initialize the count frequencies list with empty lists
        count_frequencies = [[] for i in range(len(nums) + 1)]

        # Count the frequency of each element
        for num in nums:
            number_counts[num] = number_counts.get(num, 0) + 1

        # Update the count frequencies list
        for num, count in number_counts.items():
            # Append the number to the count frequencies list at the count index
            count_frequencies[count].append(num)

        # Initialize the result list
        result = []
        # Iterate through the count frequencies list in reverse order
        for i in range(len(count_frequencies) - 1, 0, -1):
            # Append the numbers to the result list
            for num in count_frequencies[i]:
                result.append(num)
                # Break the loop if the result list has k elements
                if len(result) == k:
                    return result

        return result
# Test Cases
test = Solution()
print(test.topKFrequentBruteForce([1, 1, 1, 2, 2, 5], 2))
print(test.kTopFrequentOptimal([4, 1, -1, 2, -1, 2, 3], 2))